[WIP] mingw-w64 for haskell cross compilation with ghc.

angerman opened the pull request:
###### Motivation for this change

This includes all the necessary changes to nixpkgs, that allow me to cross compile haskell to
windows via mingw-w64.  It is not yet sufficient to allow cross compilation of Template Haskell.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

angerman committed:
check if stdenv isDarwin, not if targetPlatform isDarwin

this will break when cross compiling where targetPlatform might be something that's not darwin, however the build host is actually darwin.

angerman committed:
Drop targetPlatform == hostPlatform assert

This is triggered during the check validity phase when cross compiling.

When it is perfectly fine to use the binary package on the build machine.

angerman committed:
use custom package database

Adapted from https://github.com/obsidiansystems/nixpkgs/commit/b69f420121120433220c568e4b35ade539ef60f2 by @Ericson2314

angerman committed:
parameterize buildMK; allow ghcFlavour to be overridden.

angerman committed:
Hardcode the sourceforge mirror, while sourceforge is down.

angerman committed:
Fix /bin/haddock not found.

Only paxmark `haddock` if it was built (not all configuration build haddock).

angerman committed:
import m4

angerman committed:
Adds `doVerbose` to the generic haskell builder

angerman committed:
Doe the `links` dance only if shared is enabled.

angerman committed:
Adds buildFlags logic

angerman commented:
/cc @Ericson2314

angerman commented on pkgs/development/compilers/ghc/8.2.1-binary.nix:
```diff
@@ -4,7 +4,7 @@
 
 # Prebuilt only does native
-assert stdenv.targetPlatform == stdenv.hostPlatform;
+# assert stdenv.targetPlatform == stdenv.hostPlatform;
```
While disabling this check, allows most of the cross compilation to succeed on macOS, I believe this is also the reason why we end up with mixed (build and target) packages in the stdenv eventually.

angerman commented on pkgs/development/compilers/ghc/head.nix:
```diff
@@ -161,7 +175,11 @@ stdenv.mkDerivation rec {
   # zsh and other shells are smart about `{ghc}` but bash isn't, and doesn't
   # treat that as a unary `{x,y,z,..}` repetition.
   postInstall = ''
-    paxmark m $out/lib/${name}/bin/${if targetPlatform != hostPlatform then "ghc" else "{ghc,haddock}"}
```
This change is motivated by the fact that we do not necessarily build `haddock` with quick flavours. As such we fall over trying to `paxmark` a file that does not exist.

dtzWill replied:
Sounds good-- probably can collapse this to just `{ghc,haddock}` if we check for existence anyway.

angerman commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -116,13 +118,17 @@ let
     "--with-strip=${stdenv.cc.bintools.targetPrefix}strip"
-  ] ++ (if isHaLVM then [] else ["--hsc2hs-options=--cross-compile"]);
+  ] ++ (if isHaLVM then [] else ["--hsc2hs-option=--cross-compile"
+                                 "--hsc2hs-option=--via-asm"]);
```
`--via-asm` is a rather recent feature in `hsc2hs` that is part of ghc > 8.4

angerman commented on pkgs/os-specific/darwin/cctools/port.nix:
```diff
@@ -12,7 +12,7 @@ let
 in
 
-assert targetPlatform.isDarwin;
+assert stdenv.isDarwin;
```
This could potentially also be the reason for the incorrect package set on darwin.

angerman committed:
cleanup

angerman committed:
Revert "[cctools] check if stdenv isDarwin, not if targetPlatform isDarwin"

This reverts commit 4b6c9232b6e6b8fcfc27bf601bbeb9e3aea91818.

angerman committed:
Revert "[ghc] Drop targetPlatform == hostPlatform assert"

This reverts commit 4dac9189b82fefe0176df9b8bc98f549474e1136.

Ericson2314 commented on pkgs/development/compilers/ghc/head.nix:
```diff
@@ -15,16 +15,26 @@
 , version ? "8.5.20180118"
+, ghcRevision ? "e1d4140be4d2a1508015093b69e1ef53516e1eb6"
+, ghcSha256 ? "1gdcr10dd968d40qgljdwx9vfkva3yrvjm9a4nis7whaaac3ag58"
+, ghcFlavour ? ""
```
Better to do `ghcFlavour ? if targetPlatform != hostPlatform then "perf-cros" else ""`

angerman committed:
more verbose logic.

angerman committed:
stdenv.isDarwin -> hostplatform.isDarwin

dtzWill commented on pkgs/development/compilers/ghc/head.nix:
```diff
@@ -15,16 +15,26 @@
+  enableShared ? (targetPlatform == hostPlatform && !targetPlatform.isWindows)
+
+, # Whetherto build terminfo.
+  enableTerminfo ? !targetPlatform.isWindows
```
Does terminfo/ncurses not work on windows?

If it was failing to find ncurses headers/libs see the fix on my musl PR (remove attempt to specify where to find them).

angerman replied:
In the case of cross compiling to windows, we don't want terminfo, the GHC build system disables terminfoin that case anyway.

dtzWill replied:
It only disables it in some cross build "flavour"s, which we don't have to use as-is.  Works for me building cross-aarch64-musl and such.  Anyway sounds like you're saying windows in particular shouldn't use this, which I'll take your word for.

dtzWill commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -268,6 +278,28 @@ stdenv.mkDerivation ({
+          setupConfigureFlags+=" --extra-lib-dirs=$p/lib"
+        fi
+      done
+      ${nativeGhcCommand}-pkg --${packageDbFlag}="$setupPackageConfDir" recache
```
:1:

dtzWill commented:
If you're touching cctools, #37606 might have some useful bits.

peti commented on pkgs/development/compilers/ghc/head.nix:
```diff
@@ -15,16 +15,26 @@
-  enableShared ? true
+  #
+  # We don't do this by default if target != host, or if we target windows.
+  enableShared ? (targetPlatform == hostPlatform && !targetPlatform.isWindows)
```
I believe that the description of this option is actually quite incorrect and misleading. Setting this to true has no effect on whether ghc builds shared libraries or not. Rather, it affects the code that's used by `ghci` to load compiled objects for interpretation. There's a good chance that this setting is completely ignored on Windows anyway. Could you please check whether the change you've made here is actually necessary?

Ericson2314 replied:
`enable*` is about how the *current* package is linked. I think this does just that. That this also affects how GHCi works is just a side affect due to GHCi's fragile interpretation.

angerman committed:
Do not enable stm without enabling haskeline

Otherwise we'll end up with inconsistent dependencies.

- haskeline -> stm (1), as shipped with GHC.
- ... -> stm (2), as built with nix.

(1) /= (2) => cabal complains.

angerman committed:
allow all platforms.

angerman committed:
Adds custom Setup derivation.

angerman committed:
use/provide setup in package sets.

angerman committed:
use Setup derivation when cross compiling.

angerman committed:
do not add nativeGhc to the buildInputs

For cross compilation, we don't need it, it's deferred to the Setup derivation.
When ghc is th enativeGhc, we don't need to add it to the buildInputs either.

angerman committed:
no -staticlib for windows.

angerman committed:
Adds more asserts

angerman committed:
more logging.

angerman committed:
Revert incorrect naming.

angerman committed:
Use buildpackages for nix and nix-prefetch-scripts

angerman committed:
used stdenv.targetPlatform.isDarwin and not stdenv.isDarwin.

angerman committed:
Adds wine (darwin)

angerman committed:
fix python27

dtzWill commented:
Beep boop, how's this going?

angerman commented:
@dtzWill this does work *as-is* (I've used this to build `dhall-json` among others...). I did however get severely frustrated with the conditionals missing in `cabal2nix` generated cabal expressions. I did look into making `cabal2nix` generate better expressions, but learned that what I wanted was a bad fit for what cabal2nix does. And I don't want to disrupt any working solution in nix.

So I ended up writing [Cabal2Nix](angerman/Cabal2Nix), which really needs a better name. I now have [hackage.nix](angerman/hackage.nix), [stackage.nix](angerman/stackage.nix) as well as some glue code in [haskell.nix](angerman/haskell.nix).  This is pretty much work-in-progress. While it does allow me to build some trivial [hs-hello](angerman/Cabal2Nix.hs-hello) package, it is nowhere near polished yet, and will likely see some additional changes.

I believe it will provide a much better basis for cross compilation haskell packages.

Ericson2314 commented:
@angerman that sounds fascinating but I think your links are broken?

angerman commented:
@Ericson2314 I see... Here are the full links:

- https://github.com/angerman/Cabal2Nix
- https://github.com/angerman/hackage.nix
- https://github.com/angerman/stackage.nix
- https://github.com/angerman/haskell.nix
- https://github.com/angerman/Cabal2Nix.hs-hello

... thanks GitHub.

kirelagin commented on pkgs/development/haskell-modules/setup.nix:
```diff
@@ -0,0 +1,42 @@
+    # Add the setupHaskellDepends to a custom package-db,
+    # which ghc will be provided with.
+
+    for p in ${lib.escapeShellArgs setupHaskellDepends}; do
```
Here propagated setup dependencies are not taken into account. That is, your pkgdb will contain only immediate setup dependencies but their transitive dependencies will not be in it, so it won’t compile.

kirelagin commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -182,7 +189,21 @@ let
+  configureSetup = buildHaskellPackages.setup { setupHs = ''
+  import Distribution.Simple
+  main = defaultMainWithHooks autoconfUserHooks
+'';
```
`inherit setupHaskellDepends`, I guess.

kirelagin commented on pkgs/development/haskell-modules/setup.nix:
```diff
@@ -0,0 +1,42 @@
+
+  phases = [ "buildPhase" ];
+
+  buildInputs = [ ghc ] ++ setupHaskellDepends;
```
I think `ghc` should go to `nativeBuildInputs`.

angerman committed:
Drop `libiconv` on windows.

nixpkgs#37012 and  nixpkgs#37707 introduces the setup-hooks for libiconv, which inject `-liconv` into the `NIX_LDFLAGS`. This breaks horribly on windows where the linker end up having no idea how to linke `-liconv`. The configure.ac file specifically ignores libiconv on windows.

angerman committed:
WIP

cleverca22 committed:
fix lz4 cross-build

cleverca22 committed:
make rocksdb use cmake (doesnt work)

angerman committed:
fix libuv

Ericson2314 commented on pkgs/development/libraries/rocksdb/default.nix:
```diff
@@ -33,6 +33,10 @@ stdenv.mkDerivation rec {
   nativeBuildInputs = [ which perl snappy google-gflags ];
   buildInputs = [ zlib bzip2 lz4 malloc fixDarwinDylibNames ];
 
+  crossAttrs = {
```
Please don't use `crossAttrs` I want to remove it.

angerman committed:
rocksdb (don't fully work yet...)

angerman committed:
proper mingw names

angerman committed:
revert rocksdb

angerman committed:
cleanup mingw

angerman committed:
Adds preliminary gcc-ng infra.

angerman committed:
make bulding libgcc work.

angerman committed:
Revert "revert rocksdb"

This reverts commit 7f864704d698483122baf8b82934c0f032ec9af2.

angerman committed:
use gcc7-ng-msvcrt by default.

angerman committed:
bump

angerman committed:
Stop haskell from generating $out/lib.

Nix cc logic relies on passing /lib for each dependency as a library
search path.  This makes little sense for haskell packages, as their
libs are nested deeply in the /lib tree.  Also GHC will generate the
right library serach paths on its own from the package database
entries.

The larger issue though is that with one search path entry for each
library, and a similar one generated by GHC for each dependency
will eventually overflow the command line argument limits.

As such we do *not* put haskell libraries into $out anymore.
